docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,175 @@
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/third-party-api-servers.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Third Party API Servers
+
+## Goal
+
+With third party API servers, we want to enable cluster admins or users to
+expose new APIs at runtime without having to modify the core kubernetes API
+server binary.
+Anyone can write their own API server and register it with the core
+kubernetes API server. These new APIs will be seamless extension of core
+kubernetes APIs (ex: they will be operated upon via kubectl).
+
+## Non Goals:
+
+The following are related but are not the goals of this specific proposal:
+* Make it easy to write a kubernetes API server.
+
+## High Level Architecture
+
+Users can bring up their own API servers and register them with the core server
+at runtime.
+Core server will then start exposing these third party servers in the discovery
+API. kubectl will use the discovery API to connect to these servers.
+
+These API servers will have full control of their resources.
+Each server will have their own storage mechanism and will implement their own
+authorization and authentication.
+
+
+## Implementation Details
+
+### Thirdpartyapiservers Extensions Resource
+
+API server will support a non-namespaced `thirdpartyapiservers` extension resource.
+Third party servers can POST at
+`http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to register themselves.
+Clients can GET `http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to get a list of registered third party servers.
+The API definition will be as follows:
+
+```
+// ThirdPartyAPIServer represents the configuration of a third party server.
+type ThirdPartyAPIServer struct {
+  unversioned.TypeMeta
+
+  ObjectMeta
+
+  Spec ThirdPartyAPIServerSpec
+
+  Status ThirdPartyAPIServerStatus
+}
+
+// ThirdPartyAPIServerSpec is the specification of a ThirdPartyAPIServer
+// resource.
+type ThirdPartyAPIServerSpec struct {
+  // Base path of the API server.
+  BasePath string
+
+  // Path to the swagger spec.
+  SwaggerSpecPath string
```

Comment:
I really don't want a new place clients have to look, unless you have a
really good reason. I want this all presented in the same location.

On Mon, Nov 16, 2015 at 5:40 PM, Nikhil Jindal <notifications@github.com>
wrote:

> In docs/proposals/third-party-api-servers.md
> <https://github.com/kubernetes/kubernetes/pull/17191#discussion_r45012457>
> :
>
> > +
> > +  ObjectMeta
> > +
> > +  Spec ThirdPartyAPIServerSpec
> > +
> > +  Status ThirdPartyAPIServerStatus
> > +}
> > +
> > +// ThirdPartyAPIServerSpec is the specification of a ThirdPartyAPIServer
> > +// resource.
> > +type ThirdPartyAPIServerSpec struct {
> > +  // Base path of the API server.
> > +  BasePath string
> > +
> > +  // Path to the swagger spec.
> > +  SwaggerSpecPath string
>
> Yes swagger spec will most likely be hosted at a specified path (its
> /swaggerapi/ currently), but I thought it better to be explicit than
> assume to provide flexibility of changing it later.
>
> Here is how I was envisioning it:
>
> The main apiserver's response at /apis/ does not change. It continues to
> list the groupVersions that it supports. In addition there is another
> resource /extensions/v1beta1/thirdpartyapiservers (the path will change
> as per other comments regarding a different group version and better
> resource name) that gives the list of registered thirdparty api servers
> (including their basePath and swagger spec path). This is what clients use
> to discover them. That way the main server does not need to know the group
> versions that are supported by each of the third party servers. Just the
> swagger path is enough.
>
> Agree that we should add this explicitly in the doc.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17191/files#r45012457>.
>